OpenRA.Game/Graphics/Viewport.cs

```diff
@@ -105,9 +105,71 @@ public Viewport(WorldRenderer wr, Map map)
 
 		public CPos ViewToWorld(int2 view)
 		{
+			var world = worldRenderer.Viewport.ViewToWorldPx(view);
+			var map = worldRenderer.World.Map;
+			var ts = Game.ModData.Manifest.TileSize;
+			var candidates = CandidateMouseoverCells(world);
+			var tileSet = worldRenderer.World.TileSet;
+
+			foreach (var uv in candidates)
+			{
+				// Coarse filter to nearby cells
+				var p = map.CenterOfCell(uv.ToCPos(map.TileShape));
+				var s = worldRenderer.ScreenPxPosition(p);
+				if (Math.Abs(s.X - world.X) <= ts.Width && Math.Abs(s.Y - world.Y) <= ts.Height)
+				{
+					var ramp = 0;
+					if (map.Contains(uv))
+					{
+						var tile = map.MapTiles.Value[uv];
+						var ti = tileSet.GetTileInfo(tile);
+						if (ti != null)
+							ramp = ti.RampType;
+					}
+
+					var corners = map.CellCorners[ramp];
+					var pos = map.CenterOfCell(uv.ToCPos(map));
+					var screen = corners.Select(c => worldRenderer.ScreenPxPosition(pos + c)).ToArray();
+
+					if (screen.PolygonContains(world))
+						return uv.ToCPos(map);
+				}
+			}
+
+			// Mouse is not directly over a cell (perhaps on a cliff)
+			// Try and find the closest cell
+			if (candidates.Any())
+			{
+				return candidates.OrderBy(uv =>
+				{
+					var p = map.CenterOfCell(uv.ToCPos(map.TileShape));
+					var s = worldRenderer.ScreenPxPosition(p);
+					var dx = Math.Abs(s.X - world.X);
+					var dy = Math.Abs(s.Y - world.Y);
```

Comment:
There is a tradeoff between making the common case efficient and eating the perf in the worst case, or making the common case slower to save on the worst case.  I decided that the worst case (clicking on the side of a cliff) is rare enough that this way is better.